components/fxa-client/src/state_machine/mod.rs

```diff
@@ -0,0 +1,121 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! FxA state machine
+//!
+//! This presents a high-level API for logging in, logging out, dealing with authentication token issues, etc.
+
+use error_support::breadcrumb;
+
+use crate::{internal::FirefoxAccount, DeviceConfig, Error, FxaEvent, FxaState, Result};
+
+pub mod checker;
+mod display;
+mod internal_machines;
+
+/// Number of state transitions to perform before giving up and assuming the internal state machine
+/// is stuck in an infinite loop
+const MAX_INTERNAL_TRANSITIONS: usize = 20;
+
+use internal_machines::InternalStateMachine;
+use internal_machines::State as InternalState;
+
+impl FirefoxAccount {
+    /// Get the current state
+    pub fn get_state(&self) -> FxaState {
+        self.fxa_state.clone()
+    }
+
+    /// Process an event (login, logout, etc).
+    ///
+    /// On success, returns the new state.
+    /// On error, the state will remain the same.
+    pub fn process_event(&mut self, event: FxaEvent) -> Result<FxaState> {
+        match &self.fxa_state {
+            FxaState::Uninitialized => self.process_event_with_internal_state_machine(
+                &internal_machines::UninitializedStateMachine,
+                event,
+            ),
+            FxaState::Disconnected => self.process_event_with_internal_state_machine(
+                &internal_machines::DisconnectedStateMachine,
+                event,
+            ),
+            FxaState::Authenticating { .. } => self.process_event_with_internal_state_machine(
+                &internal_machines::AuthenticatingStateMachine,
+                event,
+            ),
+            FxaState::Connected => self.process_event_with_internal_state_machine(
+                &internal_machines::ConnectedStateMachine,
+                event,
+            ),
+            FxaState::AuthIssues => self.process_event_with_internal_state_machine(
+                &internal_machines::AuthIssuesStateMachine,
+                event,
+            ),
+        }
+    }
+
+    fn process_event_with_internal_state_machine<T: InternalStateMachine>(
+        &mut self,
+        state_machine: &T,
```

Comment:
nit: does this need to be a reference? the state_machines are short lived, so I would have expected us to pass ownership to this function

Reply:
I guess I was thinking of them as long-lived, but zero-sized, types.  If for some reason we wanted to store state in the state machines themselves, then we'd probably want to input a reference rather than an owned value.  That said, we're not doing that so let's remove the `&` symbol.